server/aesdsocket.c

```diff
@@ -0,0 +1,317 @@
+/**
+* File Name: aesdsocket.c
+* File Desciption: This file contains the main function for AESD Assignment 5 Part 1.
+* File Author: Gautama Gandhi
+* Reference: Linux man pages and Linux System Programming Textbook
+**/
+
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <errno.h>
+#include <syslog.h>
+#include <netdb.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <linux/fs.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <netdb.h>
+
+#define BACKLOG 10 // Setting 10 as the connection request limit while listening
+#define INITIAL_BUFFER_SIZE 1024 // Buffer size for receive and storage bufferss
+
+// Global Variables
+int socketfd; // Socket File Descriptor
+int testfile_fd; // Testfile File Descriptor
+int new_fd; // File Descriptor used to receive and send data
+char *storage_buffer; // Pointer to storage buffer
+
+
+// Signal handler for SIGINT and SIGTERM signals
+void sig_handler(int signum)
+{
+    if (signum == SIGINT)
+        syslog(LOG_DEBUG, "Caught signal SIGINT, exiting");
+    else if (signum == SIGTERM)
+        syslog(LOG_DEBUG, "Caught signal SIGTERM, exiting");
+
+    // Freeing storage buffer allocated memory 
+    free(storage_buffer);
+
+    //Closing socketfd FD
+    int status = close(socketfd);
+    if (status == -1) {
+        syslog(LOG_ERR, "Unable to close socket FD with error %d", errno);
+    }
+
+    // Closing testfile FD
+    status = close(new_fd);
+    if (status == -1) {
+        syslog(LOG_ERR, "Unable to close new_fd FD with error %d", errno);
+    }
+
+    // Closing testfile FD
+    status = close(testfile_fd);
+    if (status == -1) {
+        syslog(LOG_ERR, "Unable to close testfile FD with error %d", errno);
+    }
+
+    unlink("/var/tmp/aesdsocketdata");
```

Comment:
@GautamaGandhi are these safe to do from your signal handler?

Reply:
I wrote these functions based on reentrancy mentioned in the textbook. I think syslog_r is the reentrant version of syslog. I shall remove the log statements for the upcoming Assignment.

Reply:
@GautamaGandhi what about unlink?

Reply:
unlink is safely reentrant based on the textbook.

Reply:
Yes you are right about that.  What about free?

Reply:
Free is not reentrant. Should this be a part of clean up as a SIGTERM signal would result in process termination and it can occur at any time? Or would a better approach be to handle this in the while loop for receiving data and eliminating the need to call free in the signal handler?

Reply:
@GautamaGandhi see the notes and discussion from lecture today, I suggest handling this in main()